### PR TITLE
[WIP] Switch to glyphon/cosmic text from wgpu-glyph/ab-glyph

### PR DESCRIPTION
**Pull request by trimental**
Exciting times. This PR switches everything over to glyphon which effects quite a few things. Some drawbacks before we get into the positives:

* Appears to be slower at least on debug builds
* Cosmic text makes it hard to set different font sizes for things that are in the same line. This effects things like `<small>`. I've just ignored this probblem
* I've completely removed the previous font cache logic we had since it would have to be completely redone anyway
* Loading time is being affected by system font loading I believe

The benefits:
* Text does seem to be rendered better then before better then before
* Emojis work
* Text selection and 'hits' are a lot simpler to implement
* We have more control over caching behavior
* Text color seems to be improved

As a side note I don't think monospace fonts are working atm and I'm unsure why.

Here's a side by side comparison:
image

### CODE REVIEW

**Comment by CosmicHorrorDev** [comment]
> GitHub Actions has encountered an internal error when running your job.

Weird CI error. Not sure what that's about

Really looking forward to this change though! I'll happily review whenever it's finished

**Comment by CosmicHorrorDev** [comment]
Just doing some quick triaging based on the in progress changes. This branch appears to fix:

- #79 
  - Text now actually renders
- #86 
  - Since this supports font fallback

Can't reproduce the original, but may also fix

- #87 
  - I'll setup a Fedora test env this weekend since that seems to be the common link
- #102 
  - May be specific to Macs? Not able to verify

Overall the font appears slightly clearer for me, and the text is much larger, but there seem to be some minor spacing issues at least with large underlined text (from `alacritty`'s README). This doesn't appear for me on the current `main`

image

---

Overall that could be a ton of bug fixes for one PR :eyes:

**Comment by CosmicHorrorDev** [comment]
I'm able to reproduce the original issue in #87 with a plain Fedora 38 setup, but it looks like this PR potentially gets further and hits a different issue from within `glyphon`

```
thread 'main' panicked at 'called Option::unwrap() on a None value', /home/test/.cargo/git/checkouts/glyphon-bce1d8fb6a1f7cb1/80e8465/src/text_render.rs:110:26
stack backtrace:
   0: rust_begin_unwind
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:114:5
   3: glyphon::text_render::TextRenderer::prepare
   4: inlyne::renderer::Renderer::redraw
   5: inlyne::Inlyne::run::{{closure}}
   6: winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run_return
   7: winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run
   8: winit::platform_impl::platform::EventLoop<T>::run
   9: winit::event_loop::EventLoop<T>::run
  10: inlyne::main
note: Some details are omitted, run with RUST_BACKTRACE=full for a verbose backtrace.
```

This is just from running an out of the box Fedora 38 Desktop install on this branch against the repo's README

**Comment by trimental** [comment]
> Overall the font appears slightly clearer for me, and the text is much larger, but there seem to be some minor spacing issues at least with large underlined text (from `alacritty`'s README). This doesn't appear for me on the current `main`

Alright that was pretty stupid I was using x_int from [LayoutGlyph](https://docs.rs/cosmic-text/0.8.0/cosmic_text/struct.LayoutGlyph.html) and converting it into a float instead of just using x which is the float version.

**Comment by CosmicHorrorDev** [comment]
Awesome! I still see some gaps in the underline, but certainly less than before

image

I don't think this should be a blocker for this PR considering all the other things that it fixes though

**Comment by trimental** [comment]
@CosmicHorrorDev probably just a float rounding error. If you could just give it a quick lookover for any obvious improvements or mistakes we can merge this. I did some performance testing and while it is slower it still runs at like 20,000fps whilst dynamically changing text. Don't think anyones going to complain.

Btw I believe that issue you were talking about might be fixed by https://github.com/RazrFalcon/fontdb/commit/6d5658db8a90db2c4c707f354a6af30d11a7268f

**Comment by CosmicHorrorDev** [comment]
I'll give it a look later today!

Huge thanks for putting in all the work for this :)

**Review by CosmicHorrorDev — COMMENTED** [review]
Code-wise everything looks great! :+1:

From messing around a bit I noticed some issues, but I think there's only one that I would consider a blocker

---

## Blocker:

- Crash when rendering images with large dimensions

**Attempting to display:**

```markdown
<img src="https://user-images.githubusercontent.com/709451/147319227-6446589c-a4d9-480d-bd5b-43037a9e56fd.png" />
```

**Crashes `inlyne` and gives the error log:**

```text
[2023-05-18T02:08:29Z ERROR wgpu::backend::direct] Handling wgpu errors as fatal by default
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_texture
      note: label = `Image Texture`
    Dimension X value 5120 exceeds the limit of 2048

', /home/wintermute/.data/cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.16.0/src/backend/direct.rs:3019:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

---

## Non-blockers:

- Already mentioned, but there are some gaps with underlined text (`alacritty`'s README)

image

- Selecting text can physically diverge from the cursor position (`bun`'s README). This seems to vary based on the text selected instead of being some linear offset


https://github.com/trimental/inlyne/assets/30302768/6dfeab55-98cd-40bf-9fd9-8cd052455e16

---

Those are all the things I noticed. We'll also be blocked on new crates.io releases until glyphon publishes a new version, but I think that's totally fine. I'd rather have these changes in early with time to sort out any bugs anyways

**Comment by coastalwhite** [comment]
It appears the gaps in the underline are font-dependent. Maybe, it would be interesting to merge multiple underlines together into a single line. There is also an issue with *right-to-left* text here, but I feel like that might be everywhere.

File: src/interpreter.rs [hunk 1]
```diff
@@ -827,7 +827,9 @@ impl TokenSink for HtmlInterpreter {
                         text = text.make_striked(true);
                     }
                     if self.state.text_options.small >= 1 {
-                        text = text.with_size(12.);
+                        self.current_textbox.font_size = 12.;
+                        //text = text.with_size(12.);
+                        //FIXME
```
[the remark this answers is not in the document]
Comment (trimental) [reply src/interpreter.rs:832]: It is. The small tag should make only the text inside small but it doesn't seem like cosmic text has a way to set or layout a line with a variable font size.

**Comment by trimental** [comment]
@CosmicHorrorDev is that large image dimensions error a bug introduced by this pr?

**Comment by trimental** [comment]
Alright the underline and selection bugs should both be fixed now. Turns out cosmic text has a function for highlighting selections anyway.

**Comment by CosmicHorrorDev** [comment]
> @CosmicHorrorDev is that large image dimensions error a bug introduced by this pr?

Yup. I can reproduce on this branch, but not on current `main` last time I checked

File: src/renderer.rs [hunk 2]
```diff
@@ -88,15 +87,12 @@ impl Renderer {
                 &wgpu::DeviceDescriptor {
                     label: None,
                     features: wgpu::Features::empty(),
-                    limits: wgpu::Limits::downlevel_webgl2_defaults()
-                        .using_resolution(adapter.limits()),
```
[the remark this answers is not in the document]
Comment (trimental) [reply src/renderer.rs:92]: Should be fixed with https://github.com/trimental/inlyne/pull/120

**Pull request merged**